design/baremetal-operator/bmh_boot_iso.md

```diff
@@ -0,0 +1,128 @@
+
+### Goals
+
+Expose the Ironic API to boot iso images via metal3.
+
+### Non-Goals
+
+This only considers the case where the redfish-virtualmedia BMC
+option is used, not any ability to boot arbitrary iso images via
+PXE.
+
+## Proposal
+
+Add a new field to the BareMetalHost spec which indicates that instead of
+deploying an image to disk, a live-iso image will be booted:
+
+```
+  spec:
+    image:
+      iso: http://1.2.3.4/image.iso
+    online: true
+```
+
+Note that in this mode, `rootDeviceHints` and `userData` will not be used
+since the image won't be written to disk, and Ironic doesn't
+support passing user-data in addition to the iso attachment at present.
+
+At some future point it would be desirable for Ironic to add the ability
+to pass userData - then a generic iso could be booted with the ability
+to define customization via the userData field in the normal way.
+
+## Design Details
+
+When this mode is selected, we need to configure Ironic to use the [ramdisk
+deploy interface](https://docs.openstack.org/ironic/stein/admin/interfaces/deploy.html#ramdisk-deploy)
+which is not currently enabled in ironic-image.
+
+We will also need to write the provided ISO URL into the
+[instance_info boot_iso field](https://docs.openstack.org/ironic/latest/admin/drivers/redfish.html#virtual-media-ramdisk)
+
+This will only work when the redfish-virtual-media boot interface is selected
+which will be the case when the `redfish-virtualmedia+http://foo` bmc address
+is specified - we need to ensure validation exists to catch invalid combinations
+
+### Risks and Mitigations
+
+- Currently there is no detachment API in ironic, so in the case where an installer
+iso is booted, we don't have a way to detach the virtualmedia and will have to
+rely e.g on efibootmgr in whatever iso gets booted to ensure the correct boot device.
+
+- The inspection of a BMH will still rely on booting IPA and there's no support for
+fast-track provisioning in this workflow (since it's booting two different iso images)
+so this adds a reboot into the process.  This can potentially be avoided in some cases
+where we can provide a status annotation on creation of the BMH such that inspection
+is not performed.
+
+- Similarly cleaning of a BMH will rely on IPA, so will require a reboot which may
+not be desirable in some situations.  This can potentially be avoided if cleaning
+is globally disabled in the ironic-image configuration via the
+[IRONIC_AUTOMATED_CLEAN](https://github.com/metal3-io/ironic-image/blob/master/configure-ironic.sh#L39)
+variable but this doesn't provide any per-host control.
+
+### Work Items
+
+- Enable ramdisk deploy interface in ironic-image
+- Add `iso` field to the BMH schema
+- Update BMO deploy logic to set the `deploy_interface` and `instance_info` appropriately
+
+### Dependencies
+
+
```

Comment:
Add the discussed future enhancements as soft dependencies?

Reply:
Yes if you can provide links to the Ironic planned work we can include them here, or I'll add links inline with the text?